### PR TITLE
Add reconnection logic

### PR DESCRIPTION
**Pull request by carl-mastrangelo**
Tests don't currently pass due to flakiness.  Running them very slowly makes them pass tho :/

### CODE REVIEW

File: core/src/main/java/io/grpc/ClientCallImpl.java [hunk 1]
```diff
@@ -85,27 +108,21 @@
...
+
   ClientCallImpl<ReqT, RespT> setUserAgent(String userAgent) {
     this.userAgent = userAgent;
     return this;
   }
 
   @Override
-  public void start(Listener<RespT> observer, Metadata.Headers headers) {
-    Preconditions.checkState(stream == null, "Already started");
+  public void start(ClientCall.Listener<RespT> observer, Metadata.Headers headers) {
```
Comment (nmittler) [inline core/src/main/java/io/grpc/ClientCallImpl.java]: nit: maybe rename `observer` to `callListener` and then below rename `listener` to `streamListener`?

File: core/src/main/java/io/grpc/ClientCallImpl.java [hunk 2]
```diff
@@ -127,49 +144,122 @@ public void start(Listener<RespT> observer, Metadata.Headers headers) {
...
+      final ClientStreamListener listener, final Metadata.Headers headers) {
+    ClientTransport transport;
     try {
-      stream = transport.newStream(method, headers, listener);
+      transport = clientTransportProvider.get();
+    } catch (RuntimeException ex) {
+      closeCallPrematurely(listener, Status.fromThrowable(ex));
+      return false;
+    }
+    if (transport == null) {
```
Comment (nmittler) [inline core/src/main/java/io/grpc/ClientCallImpl.java]: I know this code wasn't changed, but under what conditions would `transport` be null here?

File: core/src/main/java/io/grpc/ClientCallImpl.java [hunk 3]
```diff
@@ -127,49 +144,122 @@ public void start(Listener<RespT> observer, Metadata.Headers headers) {
...
+      closeCallPrematurely(listener, Status.UNAVAILABLE.withDescription("Channel is shutdown"));
+      return false;
+    }
+
+    try {
+      synchronized (this) {
```
Comment (nmittler) [inline core/src/main/java/io/grpc/ClientCallImpl.java]: why are you synchronizing here?

File: core/src/main/java/io/grpc/ClientCallImpl.java [hunk 4]
```diff
@@ -43,25 +47,44 @@
...
-  private boolean halfCloseCalled;
+  private volatile ScheduledFuture<?> retryFuture;
   private ClientTransportProvider clientTransportProvider;
   private String userAgent;
-  private ScheduledExecutorService deadlineCancellationExecutor;
+  private ScheduledExecutorService scheduledExecutor;
+  private boolean initialBackoffComplete;
+  private double previousDelayMillis = INITIAL_BACKOFF_MILLIS;
+  private State state = State.INIT;
+  private final List<Runnable> queuedCalls = new ArrayList<Runnable>();
```
Comment (nmittler) [inline core/src/main/java/io/grpc/ClientCallImpl.java]: The name `queuedCalls` doesn't seem appropriate. AFAICT it doesn't queue call creation at all. It really just seems like a queue of operations on a call.

File: core/src/main/java/io/grpc/ClientCallImpl.java [hunk 5]
```diff
@@ -127,49 +144,122 @@ public void start(Listener<RespT> observer, Metadata.Headers headers) {
...
-    Preconditions.checkState(stream != null, "Not started");
+  public void request(final int numMessages) {
+    checkState(state == State.STARTED || state == State.HALF_CLOSED,
+        "Call was either not started or canceled");
+    synchronized (this) {
+      if (stream == null) {
+        queuedCalls.add(new Runnable() {
+          @Override
+          public void run() {
+            request(numMessages);
```
Comment (nmittler) [inline core/src/main/java/io/grpc/ClientCallImpl.java]: Should this be `stream.request(numMessages)`?

File: core/src/main/java/io/grpc/ClientCallImpl.java [hunk 6]
```diff
@@ -127,49 +144,122 @@ public void start(Listener<RespT> observer, Metadata.Headers headers) {
...
-    Preconditions.checkState(!halfCloseCalled, "call already half-closed");
-    halfCloseCalled = true;
+    checkState(state == State.STARTED, "Call was either not started or already closing: %s", state);
+    state = State.HALF_CLOSED;
+    synchronized (this) {
+      if (stream == null) {
+        queuedCalls.add(new Runnable() {
+          @Override
+          public void run() {
+            halfClose();
```
Comment (nmittler) [inline core/src/main/java/io/grpc/ClientCallImpl.java]: `stream.halfClose()`?

File: core/src/main/java/io/grpc/ClientCallImpl.java [hunk 7]
```diff
@@ -127,49 +144,122 @@ public void start(Listener<RespT> observer, Metadata.Headers headers) {
...
+      closeCallPrematurely(listener, Status.UNAVAILABLE.withDescription("Channel is shutdown"));
+      return false;
+    }
+
+    try {
+      synchronized (this) {
+        stream = transport.newStream(method, headers, listener);
+      }
+      // This will typically be empty, and no calls will be added to it once stream is not null.
+      for (Runnable queuedCall : queuedCalls) {
```
Comment (nmittler) [inline core/src/main/java/io/grpc/ClientCallImpl.java]: When would this queue ever be non-empty?

File: core/src/main/java/io/grpc/ClientCallImpl.java [hunk 8]
```diff
@@ -127,49 +144,122 @@ public void start(Listener<RespT> observer, Metadata.Headers headers) {
...
     } catch (IllegalStateException ex) {
       // We can race with the transport and end up trying to use a terminated transport.
       // TODO(ejona86): Improve the API to remove the possibility of the race.
+      retryFuture = scheduledExecutor.schedule(new Runnable() {
+        @Override
+        public void run() {
+          startInternal(listener, headers);
+        }
+      }, (long) getNextDelayMillis(), TimeUnit.MILLISECONDS);
+
```
Comment (nmittler) [inline core/src/main/java/io/grpc/ClientCallImpl.java]: It looks like you're still calling `closeCallPrematurely` below ... do you want to do that?  It doesn't seem like we should be closing anything here.

File: core/src/main/java/io/grpc/ClientCallImpl.java [hunk 9]
```diff
@@ -43,25 +47,44 @@
...
+  private static final double MAX_BACKOFF_MILLIS = 2 * 60 * 1000;
+  private static final double MIN_CONNECT_TIMEOUT_MILLIS = 20 * 1000;
+  private static final double JITTER = .2;
+
   private final MethodDescriptor<ReqT, RespT> method;
   private final SerializingExecutor callExecutor;
   private final boolean unaryRequest;
   private final CallOptions callOptions;
+  private ClientCall.Listener<?> callListener;
+  @GuardedBy("this")
```
Comment (nmittler) [inline core/src/main/java/io/grpc/ClientCallImpl.java]: Maybe guard with a separate lock object? Probably not a good idea to synchronize on an object that is accessible to the application.

File: core/src/main/java/io/grpc/ClientCallImpl.java [hunk 10]
```diff
@@ -85,27 +108,21 @@
     ClientTransport get();
   }
 
+  private enum State {INIT, STARTED, HALF_CLOSED, CANCELED}
```
Comment (nmittler) [inline core/src/main/java/io/grpc/ClientCallImpl.java]: There is a design going on for a state API ... I haven't looked at it yet, but it might be worth taking a peek at.

File: core/src/main/java/io/grpc/ClientCallImpl.java [hunk 11]
```diff
@@ -204,14 +299,42 @@ private void closeCallPrematurely(ClientStreamListener listener, Status status)
...
+            stream.cancel(Status.DEADLINE_EXCEEDED);
+          } else {
+            callListener.onClose(Status.DEADLINE_EXCEEDED, new Metadata.Trailers());
+          }
+        }
       }
     }, timeoutMicros, TimeUnit.MICROSECONDS);
   }
 
+  private double getNextDelayMillis() {
```
Comment (nmittler) [inline core/src/main/java/io/grpc/ClientCallImpl.java]: Would it be worth separating this out into a separate utility? Not sure if it would be used elsewhere, but it seems pretty generic.  WDYT?

**Comment by nmittler** [comment]
@carl-mastrangelo I took a first pass.  Looks good!

File: core/src/main/java/io/grpc/ClientCallImpl.java (hunk 7 above)
Comment (carl-mastrangelo) [inline core/src/main/java/io/grpc/ClientCallImpl.java]: It could be non empty if the transport was not ready by the time one of the ClientCall interface methods was invoked.  For example, if the transport was not yet connected, and sendPayload() was invoked, the invocation is queued onto the queuedCalls.

File: core/src/main/java/io/grpc/ClientCallImpl.java (hunk 3 above)
Comment (carl-mastrangelo) [inline core/src/main/java/io/grpc/ClientCallImpl.java]: stream is access potentially by two threads: the application thread and the scheduledExecutor thread.  Because the presence or absence of the stream variable indicates that a connection was established, I need to lock access to it.

File: core/src/main/java/io/grpc/ClientCallImpl.java (hunk 2 above)
Comment (carl-mastrangelo) [inline core/src/main/java/io/grpc/ClientCallImpl.java]: Github's diffs leave something to be desired.  The very next line is:
```
closeCallPrematurely(listener, Status.UNAVAILABLE.withDescription("Channel is shutdown"));
```

File: core/src/main/java/io/grpc/ClientCallImpl.java (hunk 1 above)
Comment (carl-mastrangelo) [inline core/src/main/java/io/grpc/ClientCallImpl.java]: I had done that at one point, but the inner class ClientStreamListenerImpl also wanted to use that name.   (I agree, those names are better!)  However, based on the comments of the previous few PRs, I was intentionally trying to change as little as possible here.  (I would be delighted to make those name changes, if you are alright with it)

File: core/src/main/java/io/grpc/ClientCallImpl.java (hunk 4 above)
Comment (carl-mastrangelo) [inline core/src/main/java/io/grpc/ClientCallImpl.java]: I agree, perhaps queuedInvocations or queuedOperations?  the word call has two meanings here.

File: core/src/main/java/io/grpc/ClientCallImpl.java (hunk 5 above)
Comment (carl-mastrangelo) [inline core/src/main/java/io/grpc/ClientCallImpl.java]: At one point in the code I had done this, but figured it would be nice to get the state checking for free.

File: core/src/main/java/io/grpc/ClientCallImpl.java (hunk 8 above)
Comment (carl-mastrangelo) [inline core/src/main/java/io/grpc/ClientCallImpl.java]: D'oh, that should be fixed.

File: core/src/main/java/io/grpc/ClientCallImpl.java [hunk 12]
```diff
@@ -43,25 +47,44 @@
...
-  private boolean cancelCalled;
-  private boolean halfCloseCalled;
+  private volatile ScheduledFuture<?> retryFuture;
   private ClientTransportProvider clientTransportProvider;
   private String userAgent;
-  private ScheduledExecutorService deadlineCancellationExecutor;
+  private ScheduledExecutorService scheduledExecutor;
+  private boolean initialBackoffComplete;
+  private double previousDelayMillis = INITIAL_BACKOFF_MILLIS;
+  private State state = State.INIT;
```
Comment (ejona86) [inline core/src/main/java/io/grpc/ClientCallImpl.java]: Is this state checking change completely orthogonal to reconnecting? I can't see any new checks or usages.

File: testing/src/main/java/io/grpc/testing/TestUtils.java [hunk 13]
```diff
@@ -151,11 +153,7 @@ public static ServerInterceptor delayServerResponseInterceptor(final AtomicLong
           @Override
           public void sendPayload(RespT payload) {
             if (delayMillis.get() != 0) {
-              try {
-                Thread.sleep(delayMillis.get());
-              } catch (InterruptedException e) {
-                Throwables.propagate(e);
-              }
+              Uninterruptibles.sleepUninterruptibly(delayMillis.get(), TimeUnit.MILLISECONDS);
```
Comment (ejona86) [inline testing/src/main/java/io/grpc/testing/TestUtils.java]: Why this change? If the thread was interrupted, I would think we would want to throw.

File: core/src/main/java/io/grpc/ClientCallImpl.java (hunk 5 above)
Comment (ejona86) [inline core/src/main/java/io/grpc/ClientCallImpl.java]: Isn't the state-checking broken like that though?

```java
call.sendPayload(o);
call.halfClosed();
```

Let's say you have the above (valid) user code. If the transport is initially down, then by the time the transport is up the call is already in half-closed state. But when sendPayload's Runnable is executed, it will throw.

File: core/src/main/java/io/grpc/ClientCallImpl.java (hunk 7 above)
Comment (ejona86) [inline core/src/main/java/io/grpc/ClientCallImpl.java]: Doesn't executing of these Runnables race with new operations by the user on the call?

**Comment by ejona86** [comment]
This really needs at least one test to make sure that requests are replayed correctly.

File: core/src/main/java/io/grpc/ClientCallImpl.java (hunk 5 above)
Comment (carl-mastrangelo) [inline core/src/main/java/io/grpc/ClientCallImpl.java]: Done.

File: core/src/main/java/io/grpc/ClientCallImpl.java (hunk 6 above)
Comment (carl-mastrangelo) [inline core/src/main/java/io/grpc/ClientCallImpl.java]: Done.

File: core/src/main/java/io/grpc/ClientCallImpl.java (hunk 4 above)
Comment (carl-mastrangelo) [inline core/src/main/java/io/grpc/ClientCallImpl.java]: Also, done.

File: core/src/main/java/io/grpc/ClientCallImpl.java (hunk 10 above)
Comment (carl-mastrangelo) [inline core/src/main/java/io/grpc/ClientCallImpl.java]: I had seen it, but  was under the impression that it was for Channels, not Calls.  I had intended this enum to just be for managing internal state, rather than using multiple booleans.

File: core/src/main/java/io/grpc/ClientCallImpl.java (hunk 11 above)
Comment (carl-mastrangelo) [inline core/src/main/java/io/grpc/ClientCallImpl.java]: Perhaps some sort of RetryPolicy class?  This class is getting kind of big as is, but I don't know a better place to put it atm.

File: core/src/main/java/io/grpc/ClientCallImpl.java (hunk 10 above)
Comment (ejona86) [inline core/src/main/java/io/grpc/ClientCallImpl.java]: Yes, it is for Channels, not Calls.

Note that when I added the state checking I considered an enum. I didn't so that it was trivial to compare the logic to the documentation.

File: core/src/main/java/io/grpc/ClientCallImpl.java (hunk 12 above)
Comment (carl-mastrangelo) [inline core/src/main/java/io/grpc/ClientCallImpl.java]: The idea was to make it simpler than checking isClose or isCanceled.

File: testing/src/main/java/io/grpc/testing/TestUtils.java (hunk 13 above)
Comment (carl-mastrangelo) [inline testing/src/main/java/io/grpc/testing/TestUtils.java]: Not sure if tests use interrupted at all.  I was debugging a flaky test and made this.  I can revert it if you'd like.

File: core/src/main/java/io/grpc/ClientCallImpl.java (hunk 12 above)
Comment (ejona86) [inline core/src/main/java/io/grpc/ClientCallImpl.java]: Did you answer my question? Is it orthogonal, or does it relate to the reconnecting? I am wondering if there is something I missed.

File: testing/src/main/java/io/grpc/testing/TestUtils.java (hunk 13 above)
Comment (ejona86) [inline testing/src/main/java/io/grpc/testing/TestUtils.java]: Please do. Although we don't treat it as such, this is a public API, so you can't assume how it is used from our tests.

File: core/src/main/java/io/grpc/ClientCallImpl.java (hunk 10 above)
Comment (carl-mastrangelo) [inline core/src/main/java/io/grpc/ClientCallImpl.java]: Acknowledged.

File: testing/src/main/java/io/grpc/testing/TestUtils.java (hunk 13 above)
Comment (carl-mastrangelo) [inline testing/src/main/java/io/grpc/testing/TestUtils.java]: Done.

File: core/src/main/java/io/grpc/ClientCallImpl.java (hunk 12 above)
Comment (carl-mastrangelo) [inline core/src/main/java/io/grpc/ClientCallImpl.java]: It does not relate to reconnecting.

File: core/src/main/java/io/grpc/ClientCallImpl.java (hunk 11 above)
Comment (ejona86) [inline core/src/main/java/io/grpc/ClientCallImpl.java]: That sounds fine to me. No need for an interface, except that it would probably make testing easier. It seems the class would just need this one method and the associated state.

File: core/src/main/java/io/grpc/ClientCallImpl.java [hunk 14]
```diff
@@ -127,70 +145,148 @@ public void start(Listener<RespT> observer, Metadata.Headers headers) {
...
+      try {
+        stream = transport.newStream(method, headers, listener);
+        // This will typically be empty, and no calls will be added to it once stream is not null.
+        for (Runnable queuedCall : queuedOperations) {
+          queuedCall.run();
+        }
+      } catch (IllegalStateException ex) {
+        // We can race with the transport and end up trying to use a terminated transport.
+        // TODO(ejona86): Improve the API to remove the possibility of the race.
+        retryFuture = scheduledExecutor.schedule(new Runnable() {
```
Comment (ejona86) [inline core/src/main/java/io/grpc/ClientCallImpl.java]: This needs to be done with the lock held.

Comment (carl-mastrangelo) [inline core/src/main/java/io/grpc/ClientCallImpl.java]: The synchronized block exchanged places with the try, I believe the lock is held?

Comment (ejona86) [inline core/src/main/java/io/grpc/ClientCallImpl.java]: You're right. It is.

File: core/src/main/java/io/grpc/ClientCallImpl.java [hunk 15]
```diff
@@ -127,70 +145,148 @@ public void start(Listener<RespT> observer, Metadata.Headers headers) {
...
+  }
+
+    return true;
   }
 
   @Override
-  public void request(int numMessages) {
-    Preconditions.checkState(stream != null, "Not started");
+  public void request(final int numMessages) {
+    checkState(state == State.STARTED || state == State.HALF_CLOSED,
```
Comment (ejona86) [inline core/src/main/java/io/grpc/ClientCallImpl.java]: This is not a correct state check. You can call request after halfClosed, and it will be commonly done for server-streaming calls. If you change the state check to check for `CANCELED`, please update the ClientCall documentation.

File: core/src/main/java/io/grpc/ClientCallImpl.java (hunk 10 above)
Comment (ejona86) [inline core/src/main/java/io/grpc/ClientCallImpl.java]: Elsewhere (like the canonical status codes) we are using the spelling cancelled, not canceled. It might be a good idea to use the double ell to have a consistent spelling.

File: core/src/main/java/io/grpc/ClientCallImpl.java (hunk 15 above)
Comment (carl-mastrangelo) [inline core/src/main/java/io/grpc/ClientCallImpl.java]: WAI?  The only two states this check will fail for are INIT and CANCELED.

File: core/src/main/java/io/grpc/ClientCallImpl.java (hunk 10 above)
Comment (carl-mastrangelo) [inline core/src/main/java/io/grpc/ClientCallImpl.java]: Done.

File: core/src/main/java/io/grpc/ClientCallImpl.java (hunk 15 above)
Comment (ejona86) [inline core/src/main/java/io/grpc/ClientCallImpl.java]: Ah, I see. Please still update ClientCall documentation.

Comment (carl-mastrangelo) [inline core/src/main/java/io/grpc/ClientCallImpl.java]: Done.

File: core/src/main/java/io/grpc/ClientCallImpl.java [hunk 16]
```diff
@@ -127,70 +145,148 @@ public void start(Listener<RespT> observer, Metadata.Headers headers) {
...
+        for (Runnable queuedCall : queuedOperations) {
+          queuedCall.run();
+        }
+      } catch (IllegalStateException ex) {
+        // We can race with the transport and end up trying to use a terminated transport.
+        // TODO(ejona86): Improve the API to remove the possibility of the race.
+        retryFuture = scheduledExecutor.schedule(new Runnable() {
+          @Override
+          public void run() {
+            startInternal(listener, headers);
```
Comment (ejona86) [inline core/src/main/java/io/grpc/ClientCallImpl.java]: We need a try-catch here in order to not lose the exception. At the very least a log. We should probably catch any exceptions and cancel the call with an INTERNAL error, but I don't think you can do that today.

Comment (carl-mastrangelo) [inline core/src/main/java/io/grpc/ClientCallImpl.java]: Done.  There is a tricky part to this where startInternal could have an unexpected exception and not reschedule itself.  As long at the rpc specified a deadline I believe it will eventually finish.  Otherwise this call will just be dead.

Also, moved the queuedOperations outside of the catch to prevent unexpected ISE's from trigger a retry.  (Exceptions as flow control, yikes).

File: core/src/main/java/io/grpc/ClientCallImpl.java (hunk 12 above)
Comment (nmittler) [inline core/src/main/java/io/grpc/ClientCallImpl.java]: @carl-mastrangelo maybe just revert the state change and just focus this PR on reconnecting? I suspect it will be easier to get it through review.  You can raise a separate issue if you want to propose a way of cleaning up channel state.

File: core/src/main/java/io/grpc/ClientCallImpl.java (hunk 1 above)
Comment (nmittler) [inline core/src/main/java/io/grpc/ClientCallImpl.java]: @carl-mastrangelo Names aren't a change deal IMO, especially if they help clarify things. If you're cool with the rename, go for it! :)

File: core/src/main/java/io/grpc/ClientCallImpl.java (hunk 11 above)
Comment (nmittler) [inline core/src/main/java/io/grpc/ClientCallImpl.java]: `RetryPolicy` SGTM as well. Agreed it would just have a single method. Maybe make an interface and then have `ExponentialBackoffPolicy` as an implementation ... WDYT?

**Comment by nmittler** [comment]
@carl-mastrangelo finished another round of review.

**Comment by ejona86** [comment]
I'm thinking this PR does not do what it is trying to do. Right now, it doesn't prevent _any_ reconnection attempts or guarantee any delay between attempts. Right now it seems to try to resolve the race when creating a new stream on a closed transport, but using exponential backoff for such a solution is probably not appropriate. It seems instead `ChannelImpl` should be managing the timer for doing a reconnection attempt.

File: core/src/main/java/io/grpc/ClientCallImpl.java (hunk 12 above)
Comment (carl-mastrangelo) [inline core/src/main/java/io/grpc/ClientCallImpl.java]: It would be easier to get through review.  Done.

File: core/src/main/java/io/grpc/ClientCallImpl.java (hunk 1 above)
Comment (carl-mastrangelo) [inline core/src/main/java/io/grpc/ClientCallImpl.java]: It's so hard to tell which oneliners are a big deal and which ones aren't. :P 

Also, Done.

**Comment by carl-mastrangelo** [comment]
Right now ChannelImpl.obtainActiveTransport is somewhat of a misnomer.  The return value can be a couple different things:

1. A completely active, ready to use transport.
2. A non active, known to be broken transport but not null transport.   
3. A non active, non broken transport, but which breaks before it can be used by the caller of oAT()
4. A null Transport, indicating that the channel is shutdown, and that no further connections can be made.  

Due to the limited nature of ClientTransport, it is not possible to distinguish between case 1, 2, or 3.  There is no way to determine if the transport is willing to send and receive messages.   Since ClientCall can not block, nor throw, nor return result codes, it just buffers calls and will retry them eventually until the deadline timer fires.  How often should ClientCall retry to get a transport?  With this change, it will eventually get one, without overloading the server (though, multiple client calls would all try to reestablish, which is what you are concerned over).

If you are saying that ChannelImpl should be doing reconnect attempts, there are two possible approaches I can think of, neither which seem very palatable.  

1.  ChannelImpl has a reconnect time _in addition_ to ClientCall.  oAT() could throw some exception, or continue to return the corpse transport as it does now, until the cooldown timer fired.  ClientCall still has to have the retry logic since it cannot reject any method invocations, and has to eventually try again.

2.  Make some sort of callback provided by ClientCall to ChannelImpl, which will be invoked when an active transport is finally established.  This will require ChannelImpl to keep track of all Calls that have been created, and make sure that finished calls are cleared out, and that calls with exceeded deadlines are properly handled (since they happen on another thread).

3.  Magical third option with serendipitously solves all the problems here.  (feel free to fill in this blank)


As it stands, this code *does* do retries, just not with connection level backoff.  Calls that occur when the transport is down will continue to try until the deadline is reached, as the PR description indicates.

**Comment by ejona86** [comment]
@carl-mastrangelo, let's step back. The issue trying to be resolved here is "Implement proper reconnection" which is primarily "do exponential backoff" when you look at the spec. Anything that does not cause "exponential backoff" on a Channel-wide basis doesn't help with the issue.

This is said last, but I want to call it out first:
> As it stands, this code does do retries

Exactly. It tries to do _call retries_ instead of _channel reconnects_. And it only solves one _very minor_ case of call retry, which is when a race condition happened during a transport shutdown. Of the cases that need to be solved with call retries, the one solved here is likely the least or second-least important. Solutions to the more important retries will also likely provide an easy way to fix this less important case.

> just not with connection level backoff

So it does look like we agree, except on what the objective is.

> Due to the limited nature of ClientTransport, it is not possible to distinguish between case 1, 2, or 3.

Is distinguishing between 1, 2, and 3 necessary? I'd say not, because I think individual Calls don't need to worry about this; they use whatever transport they are returned. However, if distinguishing between those cases is necessary, then we should enhance ClientTransport.

> There is no way to determine if the transport is willing to send and receive messages.

I'd disagree, saying that a transport is expected to function any time after start() and before transportShutdown(), but maybe we have a different definition of "willing".

It is true that "things can go wrong" after start() and the transport failed doing basic things like DNS resolution or making a TCP connection, but in many ways those are "just I/O" errors as if someone tripped on the Ethernet cable after the TCP connection was established.

> Since ClientCall can not block, nor throw, nor return result codes, it just buffers calls and will retry them eventually until the deadline timer fires. How often should ClientCall retry to get a transport?

Why should `ClientCall` retry to get the transport? Why doesn't it just give up if it fails?

> With this change, it will eventually get one, without overloading the server (though, multiple client calls would all try to reestablish, which is what you are concerned over).

But "eventually getting a transport" isn't the feature. Sane reconnection logic with exponential backoff is the feature. "multiple client calls would all try to reestablish" is what we want to _prevent_, not cause. Basically, this PR does the opposite of #609, by causing more reconnect attempts instead of fewer.

File: core/src/main/java/io/grpc/ClientCallImpl.java (hunk 1 above)
Comment (ejona86) [inline core/src/main/java/io/grpc/ClientCallImpl.java]: A general question to ask that generally leads to good results:

1. Is the change necessary for this PR?

Things that can conflict with that rule:

1. Is the change a non-trivial dependent change that would be most easily reviewed in a separate PR?

"Most easily reviewed" also includes things like "lower risk" such that you don't block one set of changes that everyone wants due to another that there is more contention about.

Style changes have higher risk and lower return, and they tend to simply be noise during review, which does hinder the time it takes to go through the review process. The changing of the state checking from booleans to enums would be a riskier thing to include, especially if it is completely orthogonal. Removing Preconditions would be similar high risk and low return.

However, limited variable renaming because you are adding a new field can improve readability and can reduce confusion. But that assumes that you changed something in the PR that made the problem worse than it was previously (like adding a new class field that has same/similar name), and that the added review noise won't make things worse for the reviewer than the rename itself helps. So that means doing a rename that touches 1000 lines of code and then having a functional change also in the PR is not going to work out well, since the rename churn makes it too hard to find the actual functional change.

Things are easier for new code than changing code. For new code the author gets to choose the approach and the reviewer should really defer to your judgement unless there are strong reasons against. For existing code, the author needs to have the reason, and that reason needs to be _stronger_ than the reasons to leave the code as-is, lest you get into constant code flip-flops and churn.

**Comment by nmittler** [comment]
@carl-mastrangelo has this PR been abandoned in favor of #672? Should we close this?

**Comment by carl-mastrangelo** [comment]
The alternative implementation has been commited, so I will revert this.
